File: Integer to Roman.py
class Solution:
	def digitToRoman(self, digit, one, five, ten):
		if (digit == 0):
			return ""
		elif (digit == 1):
			return one
		elif (digit == 2):
			return one + one
		elif (digit == 3):
			return one + one + one
		elif (digit == 4):
			return one + five
		elif (digit == 5):
			return five
		elif (digit == 6):
			return five + one
		elif (digit == 7):
			return five + one + one
		elif (digit == 8):
			return five + one + one + one
		elif (digit == 9):
			return one + ten
	# @return a string
	def intToRoman(self, num):
		str = self.digitToRoman(num % 10, 'I', 'V', 'X')
		num /= 10
		str = self.digitToRoman(num % 10, 'X', 'L', 'C') + str
		num /= 10
		str = self.digitToRoman(num % 10, 'C', 'D', 'M') + str
		num /= 10
		return self.digitToRoman(num % 10, 'M', '?', '?') + str

File: Integer to Roman.java
public class Solution {
    private String digitToRoman(int digit, char one, char five, char ten) {
        StringBuilder str = new StringBuilder("");
        if (digit == 4 || digit == 9)
            str.append(one);
        if (digit >= 4 && digit <= 8)
            str.append(five);
        if (digit == 9)
            str.append(ten);
        if (digit % 5 <= 3) {
            for (int i = 0; i < digit % 5; i++) {
                str.append(one);
            }
        }
        return str.toString();
    }
    public String intToRoman(int num) {
        int d1 = num % 10;
        String s1 = digitToRoman(d1, 'I', 'V', 'X');
        num /= 10;
        int d2 = num % 10;
        String s2 = digitToRoman(d2, 'X', 'L', 'C');
        num /= 10;
        int d3 = num % 10;
        String s3 = digitToRoman(d3, 'C', 'D', 'M');
        num /= 10;
        int d4 = num % 10;
        String s4 = digitToRoman(d4, 'M', '?', '?');
        return s4 + s3 + s2 + s1;
    }
}